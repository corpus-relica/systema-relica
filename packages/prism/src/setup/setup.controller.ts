import { Controller, Get, Post, Body } from '@nestjs/common';
import { SetupService } from './setup.service';

export interface CreateUserDto {
  username: string;
  email: string;
  password: string;
}

@Controller('setup')
export class SetupController {
  constructor(private readonly setupService: SetupService) {}

  @Get('status')
  getStatus() {
    return {
      success: true,
      data: this.setupService.getSetupState(),
    };
  }

  @Post('start')
  startSetup() {
    this.setupService.startSetup();
    return {
      success: true,
      data: {
        message: 'Setup process started',
        ...this.setupService.getSetupState(),
      },
    };
  }

  @Post('create-user')
  createUser(@Body() createUserDto: CreateUserDto) {
    const { username, email, password } = createUserDto;

    // Basic validation
    if (!username || !email || !password) {
      return {
        success: false,
        error: {
          code: 'validation-error',
          type: 'input-validation',
          message: 'Username, email, and password are required',
        },
      };
    }

    // if (password !== confirmPassword) {
    //   return {
    //     success: false,
    //     error: {
    //       code: 'validation-error',
    //       type: 'input-validation',
    //       message: 'Passwords do not match',
    //     },
    //   };
    // }

    this.setupService.submitCredentials(username, email, password);

    return {
      success: true,
      data: {
        message: 'User creation initiated',
        ...this.setupService.getSetupState(),
      },
    };
  }

  @Post('debug/reset-system')
  async resetSystem() {
    console.log('ðŸš¨ DEBUG: System reset requested');
    
    try {
      const result = await this.setupService.resetSystem();
      
      return {
        success: result.success,
        message: result.message,
        errors: result.errors,
        timestamp: new Date().toISOString(),
      };
    } catch (error) {
      console.error('DEBUG: System reset failed:', error);
      return {
        success: false,
        message: `System reset failed: ${error.message}`,
        errors: [error.message],
        timestamp: new Date().toISOString(),
      };
    }
  }
}
