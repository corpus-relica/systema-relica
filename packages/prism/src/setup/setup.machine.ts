import { createActor, createMachine, assign } from 'xstate';

export interface SetupContext {
  masterUser?: string;
  errorMessage?: string;
  statusMessage: string;
  dbCheckResult?: 'empty' | 'not-empty' | 'error';
  userCredentials?: {
    username: string;
    email: string;
    password: string;
  };
  seedFiles: string[];
  processedCsvFiles: string[];
  cacheResults: {
    facts?: 'success' | 'failed';
    lineage?: 'success' | 'failed';
    subtypes?: 'success' | 'failed';
  };
}

export type SetupEvent =
  | { type: 'START_SETUP' }
  | { type: 'DB_CHECK_COMPLETE_EMPTY' }
  | { type: 'DB_CHECK_COMPLETE_NOT_EMPTY' }
  | { type: 'SUBMIT_CREDENTIALS'; data: { username: string; email:string, password: string } }
  | { type: 'USER_CREATION_SUCCESS' }
  | { type: 'USER_CREATION_ERROR'; errorMessage: string }
  | { type: 'SEEDING_COMPLETE' }
  | { type: 'SEEDING_SKIPPED' }
  | { type: 'FACTS_CACHE_COMPLETE' }
  | { type: 'LINEAGE_CACHE_COMPLETE' }
  | { type: 'SUBTYPES_CACHE_COMPLETE' }
  | { type: 'CACHE_BUILD_COMPLETE' }
  | { type: 'ERROR'; errorMessage?: string };

export const setupMachine = createMachine({
  id: 'prism-setup',
  initial: 'idle',
  context: {
    masterUser: undefined,
    errorMessage: undefined,
    statusMessage: 'Idle',
    dbCheckResult: undefined,
    userCredentials: undefined,
    seedFiles: [],
    processedCsvFiles: [],
    cacheResults: {},
  } as SetupContext,
  states: {
    idle: {
      entry: assign({
        statusMessage: 'Idle',
      }),
      on: {
        START_SETUP: {
          target: 'checking_db',
        },
      },
    },
    
    checking_db: {
      entry: assign({
        statusMessage: 'Checking database state...',
      }),
      on: {
        DB_CHECK_COMPLETE_EMPTY: {
          target: 'awaiting_user_credentials',
          actions: assign({
            dbCheckResult: 'empty',
          }),
        },
        DB_CHECK_COMPLETE_NOT_EMPTY: {
          target: 'setup_complete',
          actions: assign({
            dbCheckResult: 'not-empty',
            statusMessage: 'Database already contains data.',
          }),
        },
        ERROR: {
          target: 'error',
          actions: assign({
            dbCheckResult: 'error',
            errorMessage: ({ event }) => event.errorMessage || 'Database check failed',
          }),
        },
      },
    },
    
    awaiting_user_credentials: {
      entry: assign({
        statusMessage: 'Awaiting admin user credentials...',
        errorMessage: undefined,  // Clear any previous error message
      }),
      on: {
        SUBMIT_CREDENTIALS: {
          target: 'creating_admin_user',
          actions: assign({
            userCredentials: ({ event }) => event.data,
            errorMessage: undefined,  // Clear error when resubmitting
          }),
        },
      },
    },
    
    creating_admin_user: {
      entry: assign({
        statusMessage: 'Creating admin user...',
        errorMessage: undefined,  // Clear any previous error
      }),
      on: {
        USER_CREATION_SUCCESS: {
          target: 'seeding_db',
          actions: assign({
            masterUser: ({ context }) => context.userCredentials?.username,
          }),
        },
        USER_CREATION_ERROR: {
          target: 'awaiting_user_credentials',
          actions: assign({
            statusMessage: ({ event }) => event.errorMessage,
            // Don't set errorMessage here - that's for the error state
          }),
        },
        ERROR: {
          target: 'error',
          actions: assign({
            errorMessage: ({ event }) => event.errorMessage || 'User creation failed',
          }),
        },
      },
    },
    
    seeding_db: {
      entry: assign({
        statusMessage: 'Starting database seed...',
      }),
      on: {
        SEEDING_COMPLETE: {
          target: 'building_caches',
          actions: assign({
            statusMessage: 'DB seeding complete.',
          }),
        },
        SEEDING_SKIPPED: {
          target: 'building_caches',
          actions: assign({
            statusMessage: 'Seeding skipped (no files?)',
          }),
        },
        ERROR: {
          target: 'error',
          actions: assign({
            errorMessage: ({ event }) => event.errorMessage || 'Seeding failed',
          }),
        },
      },
    },
    
    building_caches: {
      initial: 'building_facts_cache',
      states: {
        building_facts_cache: {
          entry: assign({
            statusMessage: 'Building entity-facts cache...',
          }),
          on: {
            FACTS_CACHE_COMPLETE: {
              target: 'building_lineage_cache',
              actions: assign({
                statusMessage: '...finished building entity-facts cache.',
                cacheResults: ({ context }) => ({
                  ...context.cacheResults,
                  facts: 'success' as const,
                }),
              }),
            },
            ERROR: {
              target: '#prism-setup.error',
              actions: assign({
                errorMessage: ({ event }) => event.errorMessage || 'Facts cache build failed',
                cacheResults: ({ context }) => ({
                  ...context.cacheResults,
                  facts: 'failed' as const,
                }),
              }),
            },
          },
        },
        
        building_lineage_cache: {
          entry: assign({
            statusMessage: 'Building entity-lineage cache...',
          }),
          on: {
            LINEAGE_CACHE_COMPLETE: {
              target: 'building_subtypes_cache',
              actions: assign({
                statusMessage: '...finished building entity-lineage cache.',
                cacheResults: ({ context }) => ({
                  ...context.cacheResults,
                  lineage: 'success' as const,
                }),
              }),
            },
            ERROR: {
              target: '#prism-setup.error',
              actions: assign({
                errorMessage: ({ event }) => event.errorMessage || 'Lineage cache build failed',
                cacheResults: ({ context }) => ({
                  ...context.cacheResults,
                  lineage: 'failed' as const,
                }),
              }),
            },
          },
        },
        
        building_subtypes_cache: {
          entry: assign({
            statusMessage: 'Building entity-subtypes cache...',
          }),
          on: {
            SUBTYPES_CACHE_COMPLETE: {
              target: 'building_caches_complete',
              actions: assign({
                statusMessage: '...finished building entity-subtypes cache.',
                cacheResults: ({ context }) => ({
                  ...context.cacheResults,
                  subtypes: 'success' as const,
                }),
              }),
            },
            ERROR: {
              target: '#prism-setup.error',
              actions: assign({
                errorMessage: ({ event }) => event.errorMessage || 'Subtypes cache build failed',
                cacheResults: ({ context }) => ({
                  ...context.cacheResults,
                  subtypes: 'failed' as const,
                }),
              }),
            },
          },
        },
        
        building_caches_complete: {
          on: {
            CACHE_BUILD_COMPLETE: {
              target: '#prism-setup.setup_complete',
              actions: assign({
                statusMessage: '...finished building caches.',
              }),
            },
          },
        },
      },
    },
    
    setup_complete: {
      entry: assign({
        statusMessage: 'Setup complete.',
      }),
      type: 'final',
    },
    
    error: {
      entry: assign({
        statusMessage: ({ context }) => `Error: ${context.errorMessage}`,
      }),
      on: {
        START_SETUP: {
          target: 'idle',
          actions: assign({
            errorMessage: undefined,
            statusMessage: 'Idle',
          }),
        },
      },
    },
  },
});

export function createSetupActor() {
  return createActor(setupMachine);
}
