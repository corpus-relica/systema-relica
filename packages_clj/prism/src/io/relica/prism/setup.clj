(ns io.relica.prism.setup
  (:require [taoensso.timbre :as log]
            [io.relica.prism.db :as db]
            [io.relica.prism.xls :as xls]
            [io.relica.prism.config :as config]
            [clojure.java.io :as io]
            [clojure.string :as str]
            [next.jdbc :as jdbc]
            [next.jdbc.result-set :as rs]
            [buddy.hashers :as hashers]))

;; Setup state tracking
(defonce setup-state (atom {:stage :not-started
                           :master-user nil
                           :status "Setup has not been initiated."
                           :progress 0
                           :error nil}))

;; Setup stages
(def setup-stages [:db-check :user-setup :db-seed :cache-build :complete])

(defn get-setup-state 
  "Returns the current setup state for external APIs."
  []
  @setup-state)

(defn advance-stage!
  "Advances the setup to the next stage."
  []
  (let [current-stage (:stage @setup-state)
        current-index (when (keyword? current-stage)
                       (.indexOf setup-stages current-stage))
        next-index (inc current-index)
        next-stage (if (and (number? next-index) (< next-index (count setup-stages)))
                     (nth setup-stages next-index)
                     :complete)]
    (swap! setup-state assoc
           :stage next-stage
           :progress (if (= next-stage :complete) 100
                       (* (/ (inc next-index) (count setup-stages)) 100)))))

(defn update-status!
  "Updates the setup status message."
  [status-message]
  (swap! setup-state assoc :status status-message)
  (log/info status-message)
  ;; We'll add a call to broadcast the update via WebSocket
  ;; This will be implemented later and injected to avoid circular dependencies
  (when-let [broadcast-fn (resolve 'io.relica.prism.websocket/broadcast-setup-update)]
    (broadcast-fn)))

(defn set-error!
  "Sets an error in the setup state."
  [error-message]
  (log/error error-message)
  (swap! setup-state assoc :error error-message))

(defn start-setup-sequence!
  "Initiates the interactive setup sequence."
  []
  (log/info "Starting setup sequence...")
  (swap! setup-state assoc 
         :stage :db-check 
         :error nil
         :status "Checking database state..."
         :progress 10)
  
  (try
    (if (db/database-empty?)
      (do
        (update-status! "Database is empty. Ready for initial setup.")
        (advance-stage!)) ; Move to :user-setup
      (do
        (update-status! "Database already contains data. Setup not required.")
        (swap! setup-state assoc :stage :complete :progress 100)))
    (catch Exception e
      (set-error! (str "Error checking database state: " (.getMessage e))))))

(defn validate-credentials
  "Validates the admin user credentials."
  [username password confirm-password]
  (cond
    (str/blank? username)
    {:valid false :message "Username cannot be blank"}
    
    (< (count username) 4)
    {:valid false :message "Username must be at least 4 characters"}
    
    (str/blank? password)
    {:valid false :message "Password cannot be blank"}
    
    (< (count password) 8)
    {:valid false :message "Password must be at least 8 characters"}
    
    (not= password confirm-password)
    {:valid false :message "Passwords do not match"}
    
    :else
    {:valid true}))

(defn create-admin-user!
  "Creates the admin user with the given credentials."
  [username password]
  (log/info "Creating admin user:" username)
  (update-status! (str "Creating admin user: " username))
  
  (try
    ;; Create PostgreSQL admin user
    (let [email (str username "@relica.io") ; Generate an email based on username
          password-hash (buddy.hashers/derive password {:algorithm :bcrypt})
          jdbc-conn (jdbc/get-connection (config/db-spec))
          
          ;; Create user
          user-result (jdbc/execute-one! jdbc-conn
                     ["INSERT INTO users
                       (email, username, password_hash, is_active, is_admin, first_name, last_name)
                       VALUES (?, ?, ?, true, true, 'Admin', 'User')
                       RETURNING id, email, username, is_active, is_admin"
                      email username password-hash]
                     {:builder-fn next.jdbc.result-set/as-unqualified-maps})
          
          ;; Create a default environment
          env-result (when user-result
                       (jdbc/execute-one! jdbc-conn
                        ["INSERT INTO environments (name) 
                          VALUES ('Default Environment')
                          RETURNING id"]
                        {:builder-fn next.jdbc.result-set/as-unqualified-maps}))
          
          ;; Link user to environment
          user-env-result (when (and user-result env-result)
                            (jdbc/execute-one! jdbc-conn
                              ["INSERT INTO user_environments
                                (user_id, environment_id, is_owner, can_write)
                                VALUES (?, ?, true, true)
                                RETURNING id"
                               (:id user-result) (:id env-result)]
                              {:builder-fn next.jdbc.result-set/as-unqualified-maps}))]
      
      (log/info "Created admin user in PostgreSQL:" (dissoc user-result :password_hash))
      (when env-result (log/info "Created default environment with ID:" (:id env-result)))
      (when user-env-result (log/info "Linked user to environment"))
      
      (if (and user-result env-result user-env-result)
        (do
          (update-status! (str "Admin user " username " created successfully with default environment"))
          (swap! setup-state assoc :master-user username)
          (advance-stage!) ; Move to :db-seed
          true)
        (do
          (set-error! (str "Failed to create admin user"
                          (when (and user-result (not env-result))
                            " - environment creation failed")
                          (when (and user-result env-result (not user-env-result))
                            " - user-environment link failed")))
          false)))
    (catch Exception e
      (set-error! (str "Error creating admin user: " (.getMessage e)))
      false)))

(defn seed-database!
  "Seeds the database with initial data from XLS files."
  []
  (log/info "Starting database seeding process...")
  (update-status! "Processing seed files...")
  
  (try
    (let [csv-file-paths (xls/process-seed-directory)]
      (if (seq csv-file-paths)
        (do
          (update-status! (str "Processed " (count csv-file-paths) " seed files. Importing data..."))
          
          (let [total-files (count csv-file-paths)
                progress-increment (/ 40 total-files)] ; 40% of progress bar for this stage
            
            (doseq [[idx csv-path] (map-indexed vector csv-file-paths)
                    :let [file-name (.getName (io/file csv-path))]]
              
              (update-status! (str "Importing " file-name " (" (inc idx) "/" total-files ")..."))
              
              ;; Update progress after each file
              (swap! setup-state update :progress + progress-increment)
              
              ;; Create nodes
              (update-status! (str "Creating entities from " file-name "..."))
              (let [node-result (db/load-nodes-from-csv! file-name)]
                (when-not (:success node-result)
                  (throw (ex-info (str "Failed to load nodes from " file-name) 
                                  {:file file-name :error (:error node-result)}))))
              
              ;; Create relationships
              (update-status! (str "Creating relationships from " file-name "..."))
              (let [rel-result (db/load-relationships-from-csv! file-name)]
                (when-not (:success rel-result)
                  (throw (ex-info (str "Failed to load relationships from " file-name)
                                  {:file file-name :error (:error rel-result)}))))))
          
          (update-status! "Database seeding completed successfully")
          (advance-stage!) ; Move to :cache-build
          true)
        (do
          (update-status! "No seed files found. Database seeding skipped.")
          (log/warn "XLS processing did not generate any CSV files. Setup cannot proceed to cache building.")
          (log/warn "Setup will remain in the db-seed stage until seed files are processed correctly.")
          ;; Do NOT advance stage when no seed files are found - this is an error condition
          false)))
    (catch Exception e
      (set-error! (str "Error during database seeding: " (.getMessage e)))
      false)))

(defn build-caches!
  "Builds necessary caches for the application."
  []
  (log/info "Building database caches...")
  (update-status! "Building database caches...")
  
  (try
    ;; Execute cache building queries
    (let [subtype-cache-result (db/execute-query! 
                              "MATCH (sub:Entity)-[:role]->(:Fact {rel_type_uid: 1146})-[:role]->(super:Entity)
                               WITH sub, super
                               MERGE (sub)-[:SUBTYPE_OF]->(super)
                               RETURN count(*) as cache_count"
                              {})
          
          ;; Add more cache building queries here as needed
          ]
      
      (if (:success subtype-cache-result)
        (do
          (update-status! "Database caches built successfully")
          (advance-stage!) ; Move to :complete
          true)
        (do
          (set-error! "Failed to build database caches")
          false)))
    (catch Exception e
      (set-error! (str "Error building database caches: " (.getMessage e)))
      false)))

(defn complete-setup!
  "Marks the setup as complete."
  []
  (log/info "Setup completed successfully.")
  (update-status! "Setup completed successfully.")
  (swap! setup-state assoc :stage :complete :progress 100))

(defn handle-setup-stage!
  "Handles the current setup stage based on the state."
  []
  (let [current-stage (:stage @setup-state)]
    (case current-stage
      :not-started (start-setup-sequence!)
      :db-check (start-setup-sequence!) ; Re-run the check if needed
      :user-setup nil ; Wait for user input via API
      :db-seed (seed-database!)
      :cache-build (build-caches!)
      :complete (complete-setup!)
      ;; Default case
      (log/warn "Unknown setup stage:" current-stage))))

;; Initialize the setup
(defn init!
  "Initializes the setup module."
  []
  (log/info "Setup module initialized"))
