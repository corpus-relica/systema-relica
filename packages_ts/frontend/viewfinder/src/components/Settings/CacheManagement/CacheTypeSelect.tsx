import React, { JSX } from 'react';

interface CacheTypeSelectProps {
  disabled: boolean;
  selectedTypes: string[];
  onSelectionChange: (types: string[]) => void;
}

const CACHE_TYPES = [
  { id: 'entity_facts', label: 'Entity Facts Cache' },
  { id: 'lineage', label: 'Lineage Cache' },
  { id: 'subtypes', label: 'Subtypes Cache' }
];

const CacheTypeSelect = ({
  disabled,
  selectedTypes,
  onSelectionChange
}: CacheTypeSelectProps) => {
  const handleCheckboxChange = (typeId: string) => {
    const newSelection = selectedTypes.includes(typeId)
      ? selectedTypes.filter(id => id !== typeId)
      : [...selectedTypes, typeId];
    onSelectionChange(newSelection);
  };

  return (
    <div className="space-y-2">
      <label className="text-sm font-medium">Select Cache Types:</label>
      <div className="space-y-2">
        {CACHE_TYPES.map(({ id, label }) => (
          <label
            key={id}
            className="flex items-center space-x-2 cursor-pointer"
          >
            <input
              type="checkbox"
              checked={selectedTypes.includes(id)}
              onChange={() => handleCheckboxChange(id)}
              disabled={disabled}
              className="rounded border-gray-300"
            />
            <span>{label}</span>
          </label>
        ))}
      </div>
    </div>
  );
};

export default CacheTypeSelect;