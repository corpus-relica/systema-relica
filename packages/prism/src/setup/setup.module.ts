import { Module, forwardRef } from '@nestjs/common';
import { SetupService } from './setup.service';
import { SetupController } from './setup.controller';
import { BatchModule } from '../batch/batch.module';
import { CacheModule } from '../cache/cache.module';

@Module({
  imports: [
    forwardRef(() => BatchModule),
    forwardRef(() => CacheModule),
  ],
  providers: [SetupService],
  controllers: [SetupController],
  exports: [SetupService],
})
export class SetupModule {}