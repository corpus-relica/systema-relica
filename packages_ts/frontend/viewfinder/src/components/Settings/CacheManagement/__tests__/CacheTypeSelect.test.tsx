import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom';
import CacheTypeSelect from '../CacheTypeSelect';
import { mockCacheTypes } from '../../../../tests/setup/cacheMockData';

describe('CacheTypeSelect', () => {
  const mockOnSelectionChange = jest.fn();

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('renders all cache type options', () => {
    render(
      <CacheTypeSelect
        disabled={false}
        selectedTypes={[]}
        onSelectionChange={mockOnSelectionChange}
      />
    );

    expect(screen.getByLabelText('All Caches')).toBeInTheDocument();
    expect(screen.getByLabelText('Entity Cache')).toBeInTheDocument();
    expect(screen.getByLabelText('Relationship Cache')).toBeInTheDocument();
  });

  it('shows selected cache types as checked', () => {
    render(
      <CacheTypeSelect
        disabled={false}
        selectedTypes={['all', 'entity']}
        onSelectionChange={mockOnSelectionChange}
      />
    );

    expect(screen.getByLabelText('All Caches')).toBeChecked();
    expect(screen.getByLabelText('Entity Cache')).toBeChecked();
    expect(screen.getByLabelText('Relationship Cache')).not.toBeChecked();
  });

  it('calls onSelectionChange when checkbox is clicked', () => {
    render(
      <CacheTypeSelect
        disabled={false}
        selectedTypes={[]}
        onSelectionChange={mockOnSelectionChange}
      />
    );

    const entityCheckbox = screen.getByLabelText('Entity Cache');
    fireEvent.click(entityCheckbox);

    expect(mockOnSelectionChange).toHaveBeenCalledWith(['entity']);
  });

  it('removes type from selection when unchecked', () => {
    render(
      <CacheTypeSelect
        disabled={false}
        selectedTypes={['all', 'entity']}
        onSelectionChange={mockOnSelectionChange}
      />
    );

    const entityCheckbox = screen.getByLabelText('Entity Cache');
    fireEvent.click(entityCheckbox);

    expect(mockOnSelectionChange).toHaveBeenCalledWith(['all']);
  });

  it('disables all checkboxes when disabled prop is true', () => {
    render(
      <CacheTypeSelect
        disabled={true}
        selectedTypes={['all']}
        onSelectionChange={mockOnSelectionChange}
      />
    );

    expect(screen.getByLabelText('All Caches')).toBeDisabled();
    expect(screen.getByLabelText('Entity Cache')).toBeDisabled();
    expect(screen.getByLabelText('Relationship Cache')).toBeDisabled();
  });

  it('does not call onSelectionChange when disabled', () => {
    render(
      <CacheTypeSelect
        disabled={true}
        selectedTypes={[]}
        onSelectionChange={mockOnSelectionChange}
      />
    );

    const entityCheckbox = screen.getByLabelText('Entity Cache');
    fireEvent.click(entityCheckbox);

    expect(mockOnSelectionChange).not.toHaveBeenCalled();
  });

  it('handles selecting all cache type', () => {
    render(
      <CacheTypeSelect
        disabled={false}
        selectedTypes={['entity']}
        onSelectionChange={mockOnSelectionChange}
      />
    );

    const allCheckbox = screen.getByLabelText('All Caches');
    fireEvent.click(allCheckbox);

    // When 'all' is selected, it should clear other selections
    expect(mockOnSelectionChange).toHaveBeenCalledWith(['all']);
  });

  it('deselects all when individual cache is selected', () => {
    render(
      <CacheTypeSelect
        disabled={false}
        selectedTypes={['all']}
        onSelectionChange={mockOnSelectionChange}
      />
    );

    const entityCheckbox = screen.getByLabelText('Entity Cache');
    fireEvent.click(entityCheckbox);

    // Selecting individual cache should remove 'all'
    expect(mockOnSelectionChange).toHaveBeenCalledWith(['entity']);
  });

  it('displays cache type descriptions', () => {
    render(
      <CacheTypeSelect
        disabled={false}
        selectedTypes={[]}
        onSelectionChange={mockOnSelectionChange}
      />
    );

    expect(screen.getByText('Rebuild all cache types')).toBeInTheDocument();
    expect(screen.getByText('Rebuild entity cache only')).toBeInTheDocument();
    expect(screen.getByText('Rebuild relationship cache only')).toBeInTheDocument();
  });

  it('applies correct CSS classes when enabled', () => {
    const { container } = render(
      <CacheTypeSelect
        disabled={false}
        selectedTypes={[]}
        onSelectionChange={mockOnSelectionChange}
      />
    );

    const checkboxes = container.querySelectorAll('input[type="checkbox"]');
    checkboxes.forEach(checkbox => {
      expect(checkbox).toHaveClass('h-4', 'w-4', 'text-blue-600');
    });
  });

  it('applies correct CSS classes when disabled', () => {
    const { container } = render(
      <CacheTypeSelect
        disabled={true}
        selectedTypes={[]}
        onSelectionChange={mockOnSelectionChange}
      />
    );

    const labels = container.querySelectorAll('label');
    labels.forEach(label => {
      expect(label).toHaveClass('opacity-50', 'cursor-not-allowed');
    });
  });

  it('takes a snapshot with no selection', () => {
    const { container } = render(
      <CacheTypeSelect
        disabled={false}
        selectedTypes={[]}
        onSelectionChange={mockOnSelectionChange}
      />
    );

    expect(container).toMatchSnapshot();
  });

  it('takes a snapshot with selections', () => {
    const { container } = render(
      <CacheTypeSelect
        disabled={false}
        selectedTypes={['all', 'entity']}
        onSelectionChange={mockOnSelectionChange}
      />
    );

    expect(container).toMatchSnapshot();
  });

  it('takes a snapshot when disabled', () => {
    const { container } = render(
      <CacheTypeSelect
        disabled={true}
        selectedTypes={['entity']}
        onSelectionChange={mockOnSelectionChange}
      />
    );

    expect(container).toMatchSnapshot();
  });
});