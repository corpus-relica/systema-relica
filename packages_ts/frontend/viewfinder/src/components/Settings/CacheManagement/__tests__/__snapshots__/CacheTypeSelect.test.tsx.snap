// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CacheTypeSelect takes a snapshot when disabled 1`] = `
<div>
  <div
    class="space-y-2"
  >
    <label
      class="text-sm font-medium"
    >
      Select Cache Types:
    </label>
    <div
      class="space-y-2"
    >
      <label
        class="flex items-center space-x-2 cursor-pointer"
      >
        <input
          class="rounded border-gray-300"
          disabled=""
          type="checkbox"
        />
        <span>
          Entity Facts Cache
        </span>
      </label>
      <label
        class="flex items-center space-x-2 cursor-pointer"
      >
        <input
          class="rounded border-gray-300"
          disabled=""
          type="checkbox"
        />
        <span>
          Lineage Cache
        </span>
      </label>
      <label
        class="flex items-center space-x-2 cursor-pointer"
      >
        <input
          class="rounded border-gray-300"
          disabled=""
          type="checkbox"
        />
        <span>
          Subtypes Cache
        </span>
      </label>
    </div>
  </div>
</div>
`;

exports[`CacheTypeSelect takes a snapshot with no selection 1`] = `
<div>
  <div
    class="space-y-2"
  >
    <label
      class="text-sm font-medium"
    >
      Select Cache Types:
    </label>
    <div
      class="space-y-2"
    >
      <label
        class="flex items-center space-x-2 cursor-pointer"
      >
        <input
          class="rounded border-gray-300"
          type="checkbox"
        />
        <span>
          Entity Facts Cache
        </span>
      </label>
      <label
        class="flex items-center space-x-2 cursor-pointer"
      >
        <input
          class="rounded border-gray-300"
          type="checkbox"
        />
        <span>
          Lineage Cache
        </span>
      </label>
      <label
        class="flex items-center space-x-2 cursor-pointer"
      >
        <input
          class="rounded border-gray-300"
          type="checkbox"
        />
        <span>
          Subtypes Cache
        </span>
      </label>
    </div>
  </div>
</div>
`;

exports[`CacheTypeSelect takes a snapshot with selections 1`] = `
<div>
  <div
    class="space-y-2"
  >
    <label
      class="text-sm font-medium"
    >
      Select Cache Types:
    </label>
    <div
      class="space-y-2"
    >
      <label
        class="flex items-center space-x-2 cursor-pointer"
      >
        <input
          class="rounded border-gray-300"
          type="checkbox"
        />
        <span>
          Entity Facts Cache
        </span>
      </label>
      <label
        class="flex items-center space-x-2 cursor-pointer"
      >
        <input
          class="rounded border-gray-300"
          type="checkbox"
        />
        <span>
          Lineage Cache
        </span>
      </label>
      <label
        class="flex items-center space-x-2 cursor-pointer"
      >
        <input
          class="rounded border-gray-300"
          type="checkbox"
        />
        <span>
          Subtypes Cache
        </span>
      </label>
    </div>
  </div>
</div>
`;
